internal/pppoe/pppoe.go

```diff
@@ -0,0 +1,116 @@
+// Package pppoe creates a PPPoE session with a remote server.
+package pppoe
+
+import (
+	"context"
+	"net"
+)
+
+// Addr is a PPPoE peer address.
+type Addr struct {
+	// Interface is the name of the network interface over which the
+	// PPPoE session is running.
+	Interface string
+	// SessionID is the session identifier for the PPPoE session.
+	SessionID int
+	// ConcentratorAddr is the Ethernet address of the remote PPPoE concentrator.
+	ConcentratorAddr net.HardwareAddr
+}
+
+func (a *Addr) Network() string { return "pppoe" }
+func (a *Addr) String() string  { return a.ConcentratorAddr.String() }
+
+// Conn is a PPPoE connection.
+type Conn struct {
+	// session is both the PPPoE framer/deframer kernel object, and a
+	// socket that lets you read/write packets into that
+	// framer. Mostly we only want this so we can bind it to a PPP
+	// interface later.
+	sessionFd int
+	// discovery is a raw ethernet PacketConn that we use to speak the
+	// PPPoE discovery protocol. We use this to set up a session, and
+	// to tear it down when we close the Conn.
+	discovery net.PacketConn
+	// addr is the address of the remote PPPoE concentrator. We use it
+	// during session teardown, but mostly it's there to provide if
+	// someone asks for Conn.RemoteAddr().
+	addr Addr
+}
+
+// New creates a PPPoE Conn on the given interface.
+func New(ctx context.Context, ifName string) (ret *Conn, err error) {
```

Comment:
"ret" and "err error" stutter are both gross in godoc.

Presumably you did this for line 47, which is unfortunate that the language requires you to name things to do that, even when that's not what you want in godoc.

In these cases I often make an exported version that's pretty that just calls an internal version that has an ugly signature with stutter names.